src/util/netif.cpp

```diff
@@ -0,0 +1,275 @@
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "send() to netlink socket: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+
+    // Receive response.
+    char response[4096];
+    ssize_t recv_result;
+    do {
+        recv_result = sock.Recv(response, sizeof(response), 0);
+    } while (recv_result < 0 && (errno == EINTR || errno == EAGAIN));
+    if (recv_result < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "recv() from netlink socket: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+
+    size_t response_len = static_cast<size_t>(recv_result);
+    for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, response_len); hdr = NLMSG_NEXT(hdr, response_len)) {
+        rtmsg* r = (rtmsg*)NLMSG_DATA(hdr);
+        int remaining_len = RTM_PAYLOAD(hdr);
+
+        // Iterate over the attributes.
+        rtattr *rta_gateway = nullptr;
+        int scope_id = 0;
+        for (rtattr* attr = RTM_RTA(r); RTA_OK(attr, remaining_len); attr = RTA_NEXT(attr, remaining_len)) {
+            if (attr->rta_type == RTA_GATEWAY) {
+                rta_gateway = attr;
+            } else if (attr->rta_type == RTA_OIF) {
+                Assume(sizeof(int) == RTA_PAYLOAD(attr));
+                std::memcpy(&scope_id, RTA_DATA(attr), sizeof(scope_id));
+            }
+        }
+
+        // Found gateway?
+        if (rta_gateway != nullptr) {
+            if (network == NET_IPV4) {
+                Assume(sizeof(in_addr) == RTA_PAYLOAD(rta_gateway));
+                in_addr gw;
+                std::memcpy(&gw, RTA_DATA(rta_gateway), sizeof(gw));
+                return CNetAddr(gw);
+            } else if (network == NET_IPV6) {
+                Assume(sizeof(in6_addr) == RTA_PAYLOAD(rta_gateway));
+                in6_addr gw;
+                std::memcpy(&gw, RTA_DATA(rta_gateway), sizeof(gw));
+                return CNetAddr(gw, scope_id);
+            }
+        }
+    }
+
+    return std::nullopt;
+}
+
+#elif defined(WIN32)
+
+#include <iphlpapi.h>
+
+std::optional<CNetAddr> QueryDefaultGateway(Network network)
+{
+    NET_LUID interface_luid = {};
+    SOCKADDR_INET destination_address = {};
+    MIB_IPFORWARD_ROW2 best_route = {};
+    SOCKADDR_INET best_source_address = {};
+    DWORD best_if_idx = 0;
+    DWORD status = 0;
+
+    // Pass empty destination address of the requested type (:: or 0.0.0.0) to get interface of default route.
+    Assume(network == NET_IPV4 || network == NET_IPV6);
+    if (network == NET_IPV4) {
+        destination_address.si_family = AF_INET;
+    } else if(network == NET_IPV6) {
+        destination_address.si_family = AF_INET6;
+    } else {
+        return std::nullopt;
+    }
+
+    status = GetBestInterfaceEx((sockaddr*)&destination_address, &best_if_idx);
+    if (status != NO_ERROR) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get best interface for default route: %s\n", SysErrorString(status));
```

Comment:
Unlike the linux code, here it does matter to call `NetworkErrorString`, because it calls `Win32ErrorString`, which calls `FormatMessage(W)` as the docs recommend: https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getbestinterfaceex  

My understanding is that using `strerror_s` (called by SysErrorString) here would be wrong:  https://stackoverflow.com/a/20057368

I wonder if we can prevent doing that by accident.

Reply:
That said, this Windows code is a lot simpler than the Linux stuff above (ducks...).

Reply:
OHH, my thinking was that it's a WIN32 API function, not a network function. So i thought `SysErrorString` would be correct. But i think you're right. "SysError" is more like "posix errno emulation error". Which is not what is needed here.

Reply:
> That said, this Windows code is a lot simpler than the Linux stuff above (ducks...).

You mean that centrally-controlled proprietary OSes sometimes have well-documented API's that consider use-cases, while FOSS often uses haphazard grabbag API's that were grown in accordance with one tool (where everyone is expected to parse text output of-)... or some custom library, with equally confusing interface. you can just say that out loud here you know... 😓 
Standards like RFCs, and POSIX (ignoring everyone's custom extensions) are rare exceptions.